Add workflow concurrency control

thomasjhuang opened the pull request:
## Tracking issue
This solves #5659  and also closes out #6309. The related flytekit change is here - https://github.com/flyteorg/flytekit/pull/3267. Docs are added in this PR - https://github.com/unionai/docs/pull/417 

## Why are the changes needed?
This work is to provide workflow concurrency control at a project level. It is a barebones implementation with the guarantee that at most `x` many workflows will be running, but it cannot guarantee that `x` many workflows are always running. We can control concurrency across all versions, and users specify the controls via `LaunchPlan` instantiation:
```
concurrency_limited_lp = LaunchPlan.get_or_create(
    name="my_concurrent_lp",
    workflow=my_workflow,
    concurrency=ConcurrencyPolicy(
        max_concurrency=3,
        behavior=ConcurrencyLimitBehavior.SKIP,
    ),
)
```

## What changes were proposed in this pull request?

The primary mechanism is fairly straightforward - whenever we attempt to launch an execution, make a db query to check for running executions given the NamedEntityIdentifier triplet (project/domain/wf_name), and if the running executions is above the threshold for concurrency (`max_concurrency`) then we immediately fail to create the execution. 

## How was this patch tested?
Unit tests added to execution_manager_test.go, but namely this was internally tested at LinkedIn since we are porting this feature externally. **More testing is in progress on local sandbox.**

### Labels
- **added**: `ConcurrencyPolicy` and logic surrounding workflow concurrency management, as well as db migration to add an index on `executions` table for `execution_phase`. 

### Setup process

### Screenshots

I was able to confirm locally that we receive the correct error message when we are running multiple workflows concurrently at limits of 1 and 2. Within the same project, running a workflow without limits works. 

```
Error: rpc error: code = ResourceExhausted desc = Concurrency limit (1) reached for launch plan concurrency_
limit_1. Skipping execution.
{"json":{},"level":"error","msg":"rpc error: code = ResourceExhausted desc = Concurrency limit (1) reached f
or launch plan concurrency_limit_1. Skipping execution.","ts":"2025-07-07T23:12:38-07:00"}
```

## Check all the applicable boxes <!-- Follow the above conventions to check the box -->

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

## Related PRs
 #5659 and #6309, will also add in flytekit PR as reference.

## Docs link
- Will add in docs 
 <div id='description'>
    <a href="https://bito.ai#summarystart"></a>
<h3>Summary by Bito</h3>
This pull request introduces a feature for managing workflow concurrency at the project level, allowing users to set limits on concurrent executions via a ConcurrencyPolicy. It includes logic to check running executions, enhances logging, and updates the database schema to support these changes, addressing issues #5659 and #6309.
<!-- Disabling unit_tests and post_effort_to_review
<br>
<br>
<b>Unit tests added</b>: True
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 2
-->
</div>

Sovietaced commented:
Can you run `make generate` in `flyteidl` dir and `make lint-fix` in `flyteadmin` dir.

davidmirror-ops commented:
@thomasjhuang thanks for your contribution. Also please check out the failing DCO check.
Do you plan to set a specific phase when an execution is skipped? That could be a bit beyond the scope of this PR though but it'd be useful.
Also please add some [docs ](https://github.com/unionai/docs)so users know how to leverage this in launchplans.

kumare3 commented:
Taking a look

thomasjhuang commented:
Okay I've added a few things - docs, unit test, and ran `make generate`. `make lint-fix` didn't indicate any issues on code changes that I've introduced here. Is there a reason why the checks are all pending? Probably ready for review now. Please also check out the related flytekit change - https://github.com/flyteorg/flytekit/pull/3267. 

@davidmirror-ops @kumare3 @Sovietaced

davidmirror-ops commented:
@thomasjhuang thanks for this amazing contribution. Sorry about this but could you port the docs update to the new repo? https://github.com/unionai/docs I'll take point on updating this repo's `docs` folder to indicate it's no longer the content that ends up rendered in the docs site

Sovietaced commented on flyteidl/protos/flyteidl/admin/launch_plan.proto:
```diff
@@ -144,6 +144,22 @@ message LaunchPlanSpec {
+    ConcurrencyLimitBehavior behavior = 2;
+}
+
+enum ConcurrencyLimitBehavior {
+    // A workflow that will be skipped because the limit has been hit.
+    SKIP = 0;
```
I agree

Sovietaced commented on flyteidl/protos/flyteidl/admin/launch_plan.proto:
```diff
@@ -144,6 +144,22 @@ message LaunchPlanSpec {
     ClusterAssignment cluster_assignment = 23;
+
+    // Concurrency setting for a given workflow to determine concurrency limits and behavior when the limit is breached
+    ConcurrencyPolicy concurrency_policy = 25;
```
Should this be 24? I wonder if you have something already using 24 in your fork :P

thomasjhuang replied:
Ah good catch, changed to 24

Sovietaced reviewed (changes requested):
I'm looking at porting a version of this to our fork and one thing I'm looking at is how this affects Flyte scheduler. Flyte scheduler is hardcoded to try and retry the execution creation up to 30 times if it fails, unless the gRPC status code in the error is `codes.AlreadyExists`. 

I think the scheduler code will need to be updated to either give up on `ResourceExhausted` or  we'll need to a way to articulate this case through some other means.

thomasjhuang commented:
> @thomasjhuang thanks for this amazing contribution. Sorry about this but could you port the docs update to the new repo? https://github.com/unionai/docs I'll take point on updating this repo's `docs` folder to indicate it's no longer the content that ends up rendered in the docs site

Got it, I'll open a new PR there.

thomasjhuang commented on flyteadmin/pkg/manager/impl/execution_manager.go:
```diff
@@ -1097,6 +1099,16 @@ func (m *ExecutionManager) launchExecution(
 	}
 
+	// Check ConcurrencyPolicy for LaunchPlan, this is a means to limit the concurrency of a launch plan across versions
+	// NOTE: There's a potential race condition here. Multiple concurrent requests
```
Yea, we avoided external locking here otherwise it would become pretty complex.

thomasjhuang commented on flyteadmin/pkg/manager/impl/execution_manager.go:
```diff
@@ -1156,6 +1168,97 @@ func (m *ExecutionManager) createExecutionModel(
+	timer := metrics.ConcurrencyCheckDuration.Start(ctxForTimer)
+	defer timer.Stop()
+
+	logger.Infof(ctx, "Checking concurrency limits for launch plan %v with policy %+v", lpID, launchPlan.GetSpec().GetConcurrencyPolicy())
```
Done

thomasjhuang commented on flyteadmin/pkg/manager/impl/execution_manager.go:
```diff
@@ -1156,6 +1168,97 @@ func (m *ExecutionManager) createExecutionModel(
+	logger.Infof(ctx, "Checking concurrency limits for launch plan %v with policy %+v", lpID, launchPlan.GetSpec().GetConcurrencyPolicy())
+
+	// Active phases based on the workflow enum
+	activePhases := []string{
```
Done

thomasjhuang commented on flyteadmin/pkg/manager/impl/execution_manager.go:
```diff
@@ -1156,6 +1168,97 @@ func (m *ExecutionManager) createExecutionModel(
+		logger.Warningf(ctx, "Concurrency limit (%d) reached for launch plan %v (active: %d)", launchPlan.GetSpec().GetConcurrencyPolicy().GetMax(), lpID, count)
+		if launchPlan.GetSpec().GetConcurrencyPolicy().GetBehavior() == admin.ConcurrencyLimitBehavior_SKIP {
+			metrics.ConcurrencyLimitHits.WithLabelValues(lpProject, lpDomain, lpName).Inc()
+			logger.Infof(ctx, "Skipping execution creation for launch plan %v due to concurrency limit", lpID)
```
Done

Sovietaced commented on flyteadmin/pkg/manager/impl/execution_manager.go:
```diff
@@ -47,6 +47,11 @@ import (
 const childContainerQueueKey = "child_queue"
 
+var activeExecutionPhases = []string{
+	core.WorkflowExecution_QUEUED.String(),
```
`UNDEFINED` is the default state so it shouldn't be problematic. 

I'm specifically seeing it in our production environment where there can be a large delta of time between submitting a workflow and it getting picked up by propeller, which would put it into a queued state.

Sovietaced replied:
Alright so we added this into our environment and it did cause problems. There seem to be some cases where old workflows can be stuck in an undefined state. We're still trying to root cause.

Sovietaced commented on flyteadmin/pkg/manager/impl/execution_manager_test.go:
```diff
@@ -4088,6 +4088,290 @@ func TestListExecutions_LegacyModel(t *testing.T) {
+										foundDomain = true
+									}
+								} else if field == "phase" {
+									if phaseList, ok := gormQueryExpr.Args.([]string); ok && len(phaseList) > 0 {
```
Can this be updated to exact match the length of the phase list? Would make the unit test more robust to catching changes.

Sovietaced commented:
@EngHabu seemed happy with this. We've been using a variation of it in production for the past couple weeks so I think its safe to land this.

popojk commented:
> > I'm looking at porting a version of this to our fork and one thing I'm looking at is how this affects Flyte scheduler. Flyte scheduler is hardcoded to try and retry the execution creation up to 30 times if it fails, unless the gRPC status code in the error is `codes.AlreadyExists`.
> > I think the scheduler code will need to be updated to either give up on `ResourceExhausted` or we'll need to a way to articulate this case through richer error details.
> 
> Makes sense - this change will increase `ResourceExhausted` errors noticeably, especially if measured as a metric and it can be misleading since it isn't really an error. Some changes on flytescheduler are probably needed. Although not implemented yet, I intend to also add the execution phase `SKIPPED` so that the UI can indicate skipped executions rather than throw resource error.

Hi @thomasjhuang @Sovietaced . Since I’m currently working on some scheduler related issues, I can go ahead and open an issue for this and submit a PR.
